Framework/DataObjects/src/EventList.cpp

```diff
@@ -1567,6 +1632,93 @@ void EventList::compressEventsParallelHelper(
     out.insert(out.end(), outputs[thread].begin(), outputs[thread].end());
 }
 
+template <class T>
+inline void EventList::compressFatEventsHelper(
+    const std::vector<T> &events, std::vector<WeightedEvent> &out,
+    const double tolerance, const Types::Core::DateAndTime &timeStart,
+    const double seconds) {
+  // Clear the output. We can't know ahead of time how much space to reserve :(
+  out.clear();
+  // We will make a starting guess of 1/20th of the number of input events.
+  out.reserve(events.size() / 20);
+
+  // The last TOF to which we are comparing.
+  double lastTof = std::numeric_limits<double>::lowest();
+  // For getting an accurate average TOF
+  double totalTof = 0;
+
+  // pulsetime bin information - stored as int nanoseconds because it
+  // implementation type for DateAndTime object
+  const int64_t pulsetimeStart = timeStart.totalNanoseconds();
+  const int64_t pulsetimeDelta = static_cast<int64_t>(seconds * SEC_TO_NANO);
+
+  // pulsetime information
+  std::vector<DateAndTime> pulsetimes; // all the times for new event
+
+  // Carrying weight and error
+  double weight = 0;
+  double errorSquared = 0;
+
+  // move up to first event that has a large enough pusletime
+  auto it = events.cbegin();
+  for (; it != events.cend(); ++it) {
+    if (it->m_pulsetime >= timeStart)
+      break;
+  }
```

Comment:
I wonder if it would be more effective to use binary search in the form of `std::lower_bound`?